src/imputors.jl

```diff
@@ -8,39 +8,75 @@ implement the `impute!(imp::<MyImputor>, data::AbstractVector)` method.
 abstract type Imputor end
 
 # A couple utility methods to avoid messing up var and obs dimensions
-obsdim(imp::Imputor) = imp.vardim == 1 ? 2 : 1
-vardim(imp::Imputor) = imp.vardim
+obsdim(dims) = dims
+vardim(dims) = dims == 1 ? 2 : 1
 
-function obswise(imp::Imputor, data::AbstractMatrix)
-    return (selectdim(data, obsdim(imp), i) for i in axes(data, obsdim(imp)))
+function obswise(data::AbstractMatrix; dims=1)
+    return (selectdim(data, obsdim(dims), i) for i in axes(data, obsdim(dims)))
 end
 
-function varwise(imp::Imputor, data::AbstractMatrix)
-    return (selectdim(data, vardim(imp), i) for i in axes(data, vardim(imp)))
+function varwise(data::AbstractMatrix; dims=2)
+    return (selectdim(data, vardim(dims), i) for i in axes(data, vardim(dims)))
 end
 
-function filterobs(f::Function, imp::Imputor, data::AbstractMatrix)
-    mask = [f(x) for x in obswise(imp, data)]
-    return imp.vardim == 1 ? data[:, mask] : data[mask, :]
+function filterobs(f::Function, data::AbstractMatrix; dims=1)
+    mask = [f(x) for x in obswise(data; dims=dims)]
+    return dims == 1 ? data[mask, :] : data[:, mask]
 end
 
-function filtervars(f::Function, imp::Imputor, data::AbstractMatrix)
-    mask = [f(x) for x in varwise(imp, data)]
-    return imp.vardim == 1 ? data[mask, :] : data[:, mask]
+function filtervars(f::Function, data::AbstractMatrix; dims=2)
+    mask = [f(x) for x in varwise(data; dims=dims)]
+    return dims == 1 ? data[:, mask] : data[mask, :]
 end
 
 """
-    impute(data, imp::Imputor)
+    splitkwargs(::Type{T}, kwargs...) where T <: Imputor -> (imp, rem)
+
+Takes an Imputor type with kwargs and returns the constructed imputor and the
+unused kwargs which should be passed to the `impute!` call.
+
+NOTE: This is used by utility methods with construct and imputor and call impute in 1 call.
+"""
+function splitkwargs(::Type{T}, kwargs...) where T <: Imputor
+    rem = Dict(kwargs...)
```

Comment:
This is needed so that `kwargs` is mutable, i.e. so we can later `delete!` entries?

Reply:
That's correct. Technically, we could use `Iterators.filter`, but this seemed cleaner since we'd still need at least 1 mutable structure.